tests/proofs/dexter/dexter-spec.md

```diff
@@ -15,6 +15,89 @@ module DEXTER-SPEC
+
+  claim <k> PUSH nat X:Int ; PUSH nat Y:Int ; PUSH nat Z:Int ; MUL ; EDIV ; IF_NONE {} { DUP ; CDR ; SWAP ; CAR ; PUSH nat 0 ; DIG 2 ; COMPARE ; EQ } => . ... </k>
+        <stack> .Stack => [ bool ?_ ] ; [ nat (Z *Int Y) /Int X ] ; .Stack </stack>
+    requires X >Int 0
+     andBool Y >=Int 0
+     andBool Z >=Int 0
+```
+
+```k
+endmodule
+```
+
+## Add Liquidity
+
+Allows callers to "mint" liquidity in excahnge for tez.
+Caller gains liquidity equal to `tez * storage.lqtTotal / storage.xtzPool`.
+
+-   Storage updates:
+
+```
+Storage( lqtTotal:  LqtTotal  => LqtTotal  + lqt_minted ;
+         tokenPool: TokenPool => TokenPool + tokens_deposited ;
+         xtzPool:   XtzPool   => XtzPool   + Tezos.amount
+       )
+```
+
+-   Operations:
+
+1. call to `transfer` entrypoint of liquidity address: Send tokens from sender to self.
+2. call to `mintOrBurn` entrypoint of token address: Adds liquidity for the sender.
+
+-   Preconditions
+
+1.  the token pool is _not_ currently updating (i.e. `storage.selfIsUpdatingTokenPool = false`)
+2.  the deadline has not passed (i.e. the `Tezos.now < input.deadline`)
+3.  the token transferred is less than `input.maxTokensDeposited`
+4.  the liquidity minted is more than `input.minLqtMinted`
+5.  lqtTotal, xtzPool and tokenPool are all positive
+
+```k
+module DEXTER-ADDLIQUIDITY-SPEC
+  imports DEXTER-VERIFICATION
+```
+
+```k
+  claim <k> #runProof(IsFA2, AddLiquidity(Owner, MinLqtMinted, MaxTokensDeposited, #Timestamp(Deadline))) => . </k>
+        <stack> .Stack </stack>
+        <selfIsUpdatingTokenPool> false </selfIsUpdatingTokenPool>
+        <mynow> #Timestamp(CurrentTime) </mynow>
+        <myamount> #Mutez(Amount) </myamount>
+        <myaddr> SelfAddress </myaddr>
+        <lqtTotal> OldLqt => OldLqt +Int (Amount *Int OldLqt) /Int XtzAmount </lqtTotal>
+        <xtzPool> #Mutez(XtzAmount => XtzAmount +Int Amount) </xtzPool>
+        <tokenPool> TokenAmount => TokenAmount +Int (Amount *Int TokenAmount) /Int XtzAmount </tokenPool>
+        <tokenAddress> TokenAddress:Address </tokenAddress>
+        <lqtAddress> LqtAddress:Address </lqtAddress>
+        <senderaddr> Sender </senderaddr>
+        <nonce> #Nonce(Nonce => Nonce +Int 2) </nonce>
+        <knownaddrs> KnownAddresses </knownaddrs>
+        <paramtype> #Type(#DexterVersionSpecificParamType(IsFA2)) </paramtype>
+        <operations> _
+                  => [ Transfer_tokens Pair Sender Pair SelfAddress ((Amount *Int TokenAmount) /Int XtzAmount) #Mutez(0) TokenAddress Nonce ] ;;
+                     [ Transfer_tokens Pair ((Amount *Int OldLqt) /Int XtzAmount) Owner #Mutez(0) LqtAddress (Nonce +Int 1) ] ;;
+                     .InternalList
+        </operations>
+    requires notBool IsFA2 // TODO Handle IsFA2==true
+     andBool CurrentTime <Int Deadline
+     andBool Amount <=Int MaxTokensDeposited
+     andBool Amount      >Int 0
+     andBool XtzAmount   >Int 0
+     andBool OldLqt      >Int 0
+     andBool TokenAmount >Int 0
```

Comment:
I know these are desired conditions, but the minimum requirement is only `XtzAmount   >Int 0`, right?

Reply:
Hmm, you're probably right. These simplified reasoning (this spec has been work in progress for over a week), but I guess we should be fine with just `XtzAmount > 0`. I'll see if I can generalize.